pkg/crane/optimize.go

```diff
@@ -76,89 +77,110 @@ func Optimize(src, dst string, prioritize []string, opt ...Option) error {
 	return nil
 }
 
-func optimizeAndPushImage(desc *remote.Descriptor, dstRef name.Reference, prioritize []string, o options) error {
+func optimizeAndPushImage(desc *remote.Descriptor, dstRef name.Reference, prioritize sets.String, o options) error {
 	img, err := desc.Image()
 	if err != nil {
 		return err
 	}
 
-	oimg, err := optimizeImage(img, prioritize)
+	missing, oimg, err := optimizeImage(img, prioritize)
 	if err != nil {
 		return err
 	}
 
+	if len(missing) > 0 {
+		return fmt.Errorf("the following prioritized files were missing from image: %v", missing.List())
+	}
+
 	return remote.Write(dstRef, oimg, o.remote...)
 }
 
-func optimizeImage(img v1.Image, prioritize []string) (v1.Image, error) {
+func optimizeImage(img v1.Image, prioritize sets.String) (sets.String, v1.Image, error) {
 	cfg, err := img.ConfigFile()
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 	ocfg := cfg.DeepCopy()
 	ocfg.History = nil
 	ocfg.RootFS.DiffIDs = nil
 
 	oimg, err := mutate.ConfigFile(empty.Image, ocfg)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	layers, err := img.Layers()
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
+	missingFromImage := sets.NewString(prioritize.UnsortedList()...)
 	olayers := make([]mutate.Addendum, 0, len(layers))
 	for _, layer := range layers {
+		missingFromLayer := []string{}
 		olayer, err := tarball.LayerFromOpener(layer.Uncompressed,
 			tarball.WithEstargz,
-			tarball.WithEstargzOptions(estargz.WithPrioritizedFiles(prioritize)))
+			tarball.WithEstargzOptions(
+				estargz.WithPrioritizedFiles(prioritize.List()),
+				estargz.WithAllowPrioritizeNotFound(&missingFromLayer),
+			))
 		if err != nil {
-			return nil, err
+			return nil, nil, err
 		}
+		missingFromImage = missingFromImage.Intersection(sets.NewString(missingFromLayer...))
```

Comment:
If the presence of missing files is considered an error, why not just return an error if this set is non-empty? This makes the method a bit easier to use, since you don't have to remember to check for returned missing files.

Reply:
Because they might be missing from one image, but not the index.

We only error before pushing because it is only at that point that we have no more places we might find the file.

Reply:
Interesting. For an index, this only fails if a prioritized file is not found in _any_ of the images. That makes some sense to me, but isn't super intuitive...

Maybe worth logging a warning or requiring a flag that says "yeah I know, these files might not be in every image".

Reply:
My original change only warned for missing files and I found the output easily got lost in the push output, so I opted instead for making it fatal (and not pushing the image(s)).

I do think it would be interesting to add a verbose mode that displays:
1. Which files are found in which layers
2. Which images are missing which files

... might be useful, but this feels more diagnostic than something I'd want always-on...

TBH I don't know how common different files are for different architectures.  I know for muti-arch kaniko (b/c of how Bazel names outputs) I ended up making the key binary `/kaniko/executor_{arch}` because it was the only way to get multiple binaries out of Bazel for the same Go 🙃 

I suspect that if this takes off, then we're going to want to have significantly more diagnostic tooling around this (e.g. inspecting the TOC of a layer, which files are prioritized?)

Reply:
> I suspect that if this takes off, then we're going to want to have significantly more diagnostic tooling around this (e.g. inspecting the TOC of a layer, which files are prioritized?)

Maybe an entire `estargz` subcommand? 😉

Reply:
It's hidden, so we can always do that later 🤷